source/type/pt_data.cpp

```diff
@@ -221,6 +222,58 @@ PT_Data::get_stop_point_connection(const StopPoint& from, const StopPoint& to) c
 }
 
 
+nt::JourneyPattern* PT_Data::get_or_create_journey_pattern(const nt::JourneyPatternKey& key){
+    const auto& it = journey_patterns_pool.find(key);
+    if (it != journey_patterns_pool.cend()) {
+        return it->second;
+    }
+    auto new_jp = new JourneyPattern{};
+
+    new_jp->is_frequence = key.is_frequence;
+    new_jp->route = key.route;
+    new_jp->commercial_mode = key.commercial_mode;
+    new_jp->physical_mode = key.physical_mode;
+    new_jp->odt_properties = key.odt_properties;
+    new_jp->name = key.name;
+    new_jp->uri = key.uri + ":adapted-" + std::to_string(journey_patterns.size()) ;
+    new_jp->idx = journey_patterns.size() - 1;
+
+    journey_patterns.push_back(new_jp);
+    journey_patterns_map[new_jp->uri] = new_jp;
+    journey_patterns_pool[key] = new_jp;
+
+    auto order = std::size_t{0};
+    for (auto* stop_point: key.stop_points) {
+
+        auto new_jp_point = new nt::JourneyPatternPoint();
+
+        new_jp_point->journey_pattern = new_jp;
+        new_jp_point->stop_point = stop_point;
+        // TODO: copy LineString
+
+        new_jp_point->idx = journey_pattern_points.size();
+        new_jp_point->uri = key.uri + ":" + boost::lexical_cast<std::string>(new_jp->journey_pattern_point_list.size());
+        new_jp_point->stop_point->journey_pattern_point_list.push_back(new_jp_point);
+
+        new_jp_point->order = order++;
```

Comment:
maybe using `new_jp->journey_pattern_point_list.size()` is better than hand counting idx.